node/node.go

```diff
@@ -701,3 +709,83 @@ func (node *Node) ShutDown() {
 	fmt.Print(msg)
 	os.Exit(0)
 }
+
+func (node *Node) populateSelfAddresses(epoch *big.Int) {
+	// reset the self addresses
+	node.keysToAddrsMutex.Lock()
+	defer node.keysToAddrsMutex.Unlock()
+	node.KeysToAddrs = map[string]common.Address{}
+	node.keysToAddrsEpoch = epoch
+
+	shardID := node.Consensus.ShardID
+	shardState, err := node.Consensus.ChainReader.ReadShardState(epoch)
+	if err != nil {
+		utils.Logger().Error().Err(err).
+			Int64("epoch", epoch.Int64()).
+			Uint32("shard-id", shardID).
+			Msg("[PopulateSelfAddresses] failed to read shard")
+		return
+	}
+
+	committee, err := shardState.FindCommitteeByID(shardID)
+	if err != nil {
+		utils.Logger().Error().Err(err).
+			Int64("epoch", epoch.Int64()).
+			Uint32("shard-id", shardID).
+			Msg("[PopulateSelfAddresses] failed to find shard committee")
+		return
+	}
+
+	for _, blskey := range node.Consensus.PubKey.PublicKey {
+		blsStr := blskey.SerializeToHexStr()
+		shardkey := shard.FromLibBLSPublicKeyUnsafe(blskey)
+		if shardkey == nil {
+			utils.Logger().Error().
+				Int64("epoch", epoch.Int64()).
+				Uint32("shard-id", shardID).
+				Str("blskey", blsStr).
+				Msg("[PopulateSelfAddresses] failed to get shard key from bls key")
+			return
+		}
+		addr, err := committee.AddressForBLSKey(*shardkey)
+		if err != nil {
+			utils.Logger().Error().Err(err).
+				Int64("epoch", epoch.Int64()).
+				Uint32("shard-id", shardID).
+				Str("blskey", blsStr).
+				Msg("[PopulateSelfAddresses] could not find address")
+			return
+		}
+		node.KeysToAddrs[blsStr] = *addr
+		utils.Logger().Debug().
+			Int64("epoch", epoch.Int64()).
+			Uint32("shard-id", shardID).
+			Str("bls-key", blsStr).
+			Str("address", common2.MustAddressToBech32(*addr)).
+			Msg("[PopulateSelfAddresses]")
+	}
+}
+
+// GetAddressForBLSKey retrieves the ECDSA address associated with bls key for epoch
+func (node *Node) GetAddressForBLSKey(blskey *bls.PublicKey, epoch *big.Int) common.Address {
+	// populate if first time setting or new epoch
+	if node.keysToAddrsEpoch == nil || epoch.Cmp(node.keysToAddrsEpoch) != 0 {
```

Comment:
Should lock in this func. Reading on a object that's changed by other thread also crash.